doc/flamegraphs.md

```diff
@@ -0,0 +1,108 @@
+## Flame Graphs
+
+Flame graphs are a way of visualizing software CPU and heap profiles. Any tool
+that can create stack data in the correct format can be used to generate a flame
+graph. You can learn more about flame graphs
+[here](http://www.brendangregg.com/flamegraphs.html).
+
+Example flame graph (click to zoom):
+
+![Example Flamegraph](/contrib/flamegraphs/example.svg)
+
+### Installing The Flame Graph Scripts
+
+The upstream GitHub repo for flame graphs is at
+[brendangregg/FlameGraph](https://github.com/brendangregg/FlameGraph). To run
+the scripts you'll need a Perl installation on your system.
+
+There is a small patch in the Bitcoin repository at
+[contrib/flamegraphs/bitcoin-annotations.patch](/contrib/flamegraphs/bitcoin-annotations.patch)
+that you can apply to generate flame graphs that are color coded for Bitcoin.
+You can apply it using a command like this:
+
+```bash
+# Assuming you're at the top directory in the FlameGraph repo
+$ patch -p1 <path/to/bitcoin-annotations.patch
+```
+
+### Installing Perf
+
+In general you will need a version of the
+[perf](https://perf.wiki.kernel.org/index.php/Tutorial) command that matches the
+kernel that you are using. Perf is maintained along with the kernel source tree,
+and is available on most Linux distributions:
+
+```bash
+# Install perf on Debian.
+sudo apt-get install linux-perf
+
+# Install perf on Fedora.
+sudo dnf install perf
+```
+
+### Generating Flame Graphs On Linux Using Perf
+
+To generate flame graphs with labeled function names you'll need a build of
+Bitcoin that has debugging symbols enabled (this should be the case by default).
+You can check that your executable has debugging symbols by checking for a
+`.debug_str` section in the ELF file:, e.g.
+
+```bash
+# If this produces output your binary has debugging symbols.
+$ readelf -S src/bitcoind | grep .debug_str
```

Comment:
For OSX, use `greadelf` instead of `readelf`, which can be installed using `brew install binutils`